ADRs/2022-12-08-PropertyWrappers-ADR.md

```diff
@@ -0,0 +1,184 @@
+
+All `StateContainer` properties and functions would then be accessed directly from the `@ViewState` wrapper by way of the `_` underscore prefix.
+
+Example Usage
+
+```swift
+struct SomeView: View {
+    @ViewState var state: SomeViewState
+    ...
+    var body: some View {
+        Button("\(state.isEnabled)") {
+            _state.observe(state.toggle(isEnabled: !state.isEnabled))
+        }
+        .onChange(of: state.isEnabled) { isEnabled in 
+            print(isEnabled)
+        }
+        .onReceive(_state.publisher) { newState in 
+            print(newState)
+        }
+    }
+}
+
+// Enables simple default initializer:
+SomeView(state: .init(SomeViewState()))
+```
+
+As you can see from the example, the `ViewStateRendering` protocol is no longer necessary.
+
+Benefits
+
+- Solves all the problems that `@AutoRendered` solves
+- Works for both UIKit and SwiftUI
+- More ergonomic for declaring the view state property on a view
+- More ergonomic for instantiating a view
+- No longer requires a protocol
+- The developer no longer has to choose between `@ObservedObject` and `@StateObject` for the `container` property in SwiftUI
+
+Drawbacks
+
+- The `_` prefix is generally unpopular for such a use case, causing much nose-wrinkling
+- Extra typing (`_state...`) is required on some lines because `observe` and `bind` are no longer accessible directly on the view
+- The developer is further removed from the "metal" of the framework (the `StateContainer`)
+
+### ViewState Property Wrapper with ViewStateRendering Extension
+
+This solution extends the above "ViewStateProperty Wrapper" solution by restoring the `ViewStateRendering` protocol to access the underlying `StateContainer` members, such as `observe` and `bind` without having to go through the `_state` property.
+
+It does this by using reflection in a protocol extension to map `observe`, `bind`, et al. to the `StateContainer` found within the `@ViewState` property wrapper.
+
+Example Usage
+
+```swift
+struct SomeView: View {
+    @ViewState var state: SomeViewState
+    ...
+    var body: some View {
+        Button("\(state.isEnabled)") {
+            observe(state.toggle(isEnabled: !state.isEnabled))
+        }
+        .onChange(of: state.isEnabled) { isEnabled in 
+            print(isEnabled)
+        }
+        .onReceive(statePublisher) { newState in 
+            print(newState)
+        }
+    }
+}
+```
+
+Under the hood, the `ViewStateRendering` protocol is extended like so:
+
+```swift
+public extension ViewStateRendering {
+    var container: StateContainer<State> {
+        let mirror = Mirror(reflecting: self)
+        let viewStatePropertyWrapper = mirror
+            .children
+            .filter({ $0.label == "_state" })
+            .first?.value as? ViewState<State>
+        return viewStatePropertyWrapper?.container ?? StateContainer(state: state)
```

Comment:
This feels counter to the type safety that VSM advocates.

Reply:
While I agree that VSM advocates for type safety with respect to how the pattern is consumed by implementers, when it comes to internal implementation code, it follows more of a "Type safety for thee but not for me" paradigm, where internal assumptions can be made and unit tested internally with acceptable failovers. The type safety concerns should never spill over into the framework ergonomics/API.

That being said, I agree that this is borderline because it *does have* some downstream implications for the implementer, which is one of the reasons why it wasn't promoted as the proposed solution. The counter argument being: This is *probably* no less safe than forgetting to use `@StateObject` or `@ObservedObject` in a SwiftUI view.